data-processing/scripts/convert_colossal_oscar_10_to_jsonlgz.py

```diff
@@ -0,0 +1,130 @@
+    "created": "..."         # OPTIONAL: timestamp when orig document was created (best-guess if not available), should be specified as a range; "YYYY-MM-DDTHH:MM:SS.TIMEZONE, YYYY-MM-DDTHH:MM:SS.TIMEZONE"
+    "metadata": {            # OPTIONAL: source-specific metadata
+        "sub-source": "...", # OPTIONAL: E.g. "newspaper_ocr"
+        ...
+    }
+}
+"""
+import datetime
+import gzip
+import io
+import json
+import multiprocessing
+import sys
+from pathlib import Path
+from typing import Union, Any
+
+import zstandard
+
+DOCS_PER_FILE = 1000000
+
+def process_one(output_dir: Path, root_path: Path, input_path: Path) -> None:
+    doc_count = 0
+    out_fh = None
+    with input_path.open("rb") as in_fh:
+        dctx = zstandard.ZstdDecompressor(max_window_size=2147483648)
+        stream_reader = dctx.stream_reader(in_fh)
+        text_stream = io.TextIOWrapper(stream_reader, encoding="utf-8")
+        for line in text_stream:
+            if (doc_count % DOCS_PER_FILE) == 0:
+                # Convert relative path colossal-oscar-1.0/11-17/da_meta/da_meta_part_2.jsonl.zst
+                # to output_path/11-17__da_meta__da_meta_part2_{idx}.jsonl.gz
+                rel_input_path_parts = input_path.relative_to(root_path).parts
+                output_path = output_dir / ("__".join(rel_input_path_parts[:-1]) + "__" + 
+                    Path(Path(rel_input_path_parts[-1]).stem).stem + f"_{doc_count//DOCS_PER_FILE}.jsonl.gz"
+                )
+                if out_fh is not None:
+                    out_fh.close()
+                out_fh = gzip.open(output_path, "wt")
+            obj = json.loads(line)
+            new_obj: dict[str, Union[dict[str, Any], str]] = {
+                "id": str(obj["warc_headers"]["warc-record-id"]), # Copy metadata id to root
+                "text": obj["content"],
+                "source": "colossal_oscar_1.0",
+                "added": datetime.datetime.now(datetime.UTC).strftime("%Y-%m-%dT%H:%M:%S.000Z"),
+                "created": obj["warc_headers"]["warc-date"], # Copy metadata to root
+                "metadata": obj["metadata"],
+            }
+            new_obj["metadata"]["warc_headers"] = obj["warc_headers"] # type: ignore
+            new_obj["metadata"]["url"] = obj["warc_headers"]["warc-target-uri"] # type: ignore
+
+            assert out_fh is not None
+            json.dump(new_obj, out_fh)
+            doc_count += 1
+            out_fh.write("\n")
+
+def main():
+    assert (
+        len(sys.argv) == 3
+    ), "Usage: convert_colossal_oscar_10_to_jsonlgz.py colossal_oscar_root output_directory"
+
+    oscar_root_dir = Path(sys.argv[1])
+    output_dir = Path(sys.argv[-1])
+
+    lang_codes = ["da", "sv", "no", "nn", "is"]
+    for lang in lang_codes:
+        input_files = oscar_root_dir.glob(f"./**/{lang}_meta/*.jsonl.zst")
```

Comment:
zst? Shouldn't we keep it all as json.gz?

Reply:
zst is only for the input (raw) dataset files. The output of the script is saved in jsonl.gz